src/main/scala/org/deepdive/inference/datastore/SQLInferenceDataStore.scala

```diff
@@ -656,41 +432,102 @@ trait SQLInferenceDataStore extends InferenceDataStore with Logging {
-      """)
+  // ground factor meta data
+  def groundFactorMeta(du: DataLoader, factorDescs: Seq[FactorDesc], groundingPath: String, 
+    parallelGrounding: Boolean) {
+    ds.dropAndCreateTable(FactorMetaTable, "name text, funcid int, sign text")
 
     // generate a string containing the signs (whether negated) of variables for each factor
     factorDescs.foreach { factorDesc =>
       val signString = factorDesc.func.variables.map(v => !v.isNegated).mkString(" ")
-      val funcid = getFactorFunctionTypeid(factorDesc.func.getClass.getSimpleName)
+      val funcid = InferenceNamespace.getFactorFunctionTypeid(factorDesc.func.getClass.getSimpleName)
       execute(s"INSERT INTO ${FactorMetaTable} VALUES ('${factorDesc.name}', ${funcid}, '${signString}')")
     }
 
     // dump factor meta data
-    du.unload(s"dd_factormeta", s"${groundingPath}/dd_factormeta", dbSettings, parallelGrounding,
-      s"SELECT * FROM ${FactorMetaTable}")
+    val groundingDir = getFileNameFromPath(groundingPath)
+    du.unload(InferenceNamespace.getFactorMetaFileName, 
+      s"${groundingPath}/${InferenceNamespace.getFactorMetaFileName}", 
+      dbSettings, parallelGrounding, s"SELECT * FROM ${FactorMetaTable}",
+      groundingDir)
+  }
+
+  // create feature stats for boolean LR function
+  def createFeatureStats(factorDesc: FactorDesc, querytable: String, weightlist: String,
+    weightDesc: String) {
+    // Create feature statistics support tables for error analysis, 
+    // only if it's boolean LR feature (the most common one)
+    if (factorDesc.func.variables.length == 1 && factorDesc.func.variableDataType == "Boolean") {
+      // This should be a single variable, e.g. "is_true"
+      val variableName = factorDesc.func.variables.map(v => 
+          s""" ${quoteColumn(v.toString)} """).mkString(",")
+      val groupByClause = weightlist match {
+        case "" => ""
+        case _ => s"GROUP BY ${weightlist}"
+      }
+      execute(s"""
+      INSERT INTO ${FeatureStatsSupportTable}
+      SELECT ${weightDesc} as description,
+             count(CASE WHEN ${variableName}=TRUE THEN 1 ELSE NULL END) AS pos_examples,
+             count(CASE WHEN ${variableName}=FALSE THEN 1 ELSE NULL END) AS neg_examples,
+             count(CASE WHEN ${variableName} IS NULL THEN 1 ELSE NULL END) AS queries
+      FROM ${querytable}
+      ${groupByClause};
+      """)
+      execute(analyzeTable(FeatureStatsSupportTable))
+    }
+  }
+
+  // convert grounding file format to be compatible with sampler
+  // for more information about format, please refer to deepdive's website
+  def convertGroundingFormat(groundingPath: String) {
+    log.info("Converting grounding file format...")
+    val ossuffix = if (System.getProperty("os.name").startsWith("Linux")) "linux" else "mac"
+    // TODO: this python script is dangerous and ugly. It changes too many states!
+    val cmd = s"python ${Context.deepdiveHome}/util/tobinary.py ${groundingPath} " + 
```

Comment:
Use Java File object to build paths like this, not manual "/" chars and hard-coded strings.  File will take care of directory separators and control characters, etc.  Also, "util" and "tobinary.py" should be reserved constants